train/utils.py

```diff
@@ -60,167 +58,169 @@ def to_tensor(array):
 
 
 def to_numpy(tensor):
+    if tensor is None:
+        return None
     return tensor.detach().cpu().numpy() if isinstance(tensor, torch.Tensor) else tensor
 
 
-def weighted_soft_accuracy(
-    y_pred: torch.Tensor,
-    y_true: torch.Tensor,
-    weight_positive: float = 0.95,
-    weight_negative: float = 0.05,
-    tolerance: float = 0.1,
-) -> float:
+def soft_continuous_accuracy(y_pred: torch.Tensor, y_true: torch.Tensor) -> float:
     """
-    Accuratezza pesata: le predizioni corrette per classi positive (1) valgono di più (0.95),
-    mentre quelle per classi negative (0) valgono meno (0.05).
-    E' intesa per essere usata come debug per vedere se la rete sta effettivamente migliorando
-    Non è intesa per essere usata come metrica di valutazione finale.
+    Accuracy continua: 1 - |pred - true| mediato.
+    Valori predetti vicini al target sono premiati di più.
     """
     with torch.no_grad():
-        diff = torch.abs(y_pred - y_true)
-        correct = (diff <= tolerance).float()
+        error = torch.abs(y_pred - y_true)
+        score = 1.0 - error
+        return score.mean().item()
 
-        # Assegna pesi: 0.95 per le note attive, 0.05 per quelle inattive
-        weights = torch.where(y_true > 0.5, weight_positive, weight_negative)
-        weighted_correct = correct * weights
 
-        return weighted_correct.sum().item() / weights.sum().item()
+def binary_classification_metrics(
+    y_pred: torch.Tensor, y_true: torch.Tensor, threshold: float = 0.5
+):
+    with torch.no_grad():
+        y_pred_bin = (y_pred >= threshold).float()
+
+        tp = (y_pred_bin * y_true).sum().item()
+        fp = (y_pred_bin * (1 - y_true)).sum().item()
+        fn = ((1 - y_pred_bin) * y_true).sum().item()
+        tn = ((1 - y_pred_bin) * (1 - y_true)).sum().item()
+
+        precision = tp / (tp + fp + 1e-8)
+        recall = tp / (tp + fn + 1e-8)
+        f1 = 2 * precision * recall / (precision + recall + 1e-8)
 
+        accuracy = soft_continuous_accuracy(y_pred, y_true)
 
-def plot_prediction_vs_ground_truth(yo_pred, yp_pred, yn_pred, yo_true, yn_true):
+        return {
+            "TP": tp,
+            "FP": fp,
+            "FN": fn,
+            "TN": tn,
+            "Precision": precision,
+            "Recall": recall,
+            "F1": f1,
+            "Accuracy": accuracy,
+        }
+
+
+def plot_prediction_vs_ground_truth(
+    yo_pred, yp_pred, yn_pred, yo_true, yp_true, yn_true
+):
```

Comment:
Se non è un problema, modificherei l'immagine per avere solo uno dei due (ground truth/output)
Così riusciamo a sparare su wandb il ground truth solo una volta, e l'output tante volte, poi li affianchiamo dalla dashboard.
In questo modo penso che dovremmo usare meno disco su wandb, che ne abbiamo abbastanza ma non infinito (e mi piacerebbe anche loggare audio e midi, che vanno a fare altri dati)